sdk/include/platform/sunburst/platform-ethernet.hh

```diff
@@ -0,0 +1,662 @@
+		register_write(RegisterOffset::TXCR, 0x00EE);
+		// Enable QMU Receive Frame Data Pointer Auto Increment.
+		register_write(RegisterOffset::RXFDPR, 0x4000);
+		// Configure Receive Frame Threshold for one frame.
+		register_write(RegisterOffset::RXFCTR, 0x0001);
+		// Enable QMU Receive flow control / Receive all broadcast frames
+		// /Receive unicast frames, and IP/TCP/UDP checksum verification etc.
+		register_write(RegisterOffset::RXCR1, 0x7CE0);
+		// Enable QMU Receive UDP Lite frame checksum verification, UDP Lite
+		// frame checksum generation, IPv6 UDP fragment frame pass, and
+		// IPv4/IPv6 UDP UDP checksum field is zero pass. In addition (not in
+		// the programmer's guide), enable single-frame data burst.
+		register_write(RegisterOffset::RXCR2, 0x009C);
+		// Enable QMU Receive Frame Count Threshold / RXQ Auto-Dequeue frame.
+		register_write(RegisterOffset::RXQCR,
+		               RxQueueControl::RxFrameCountThresholdEnable |
+		                 RxQueueControl::AutoDequeueRxQFrameEnable);
+
+		// Programmer's guide have a step to set the chip in half-duplex when
+		// negotiation failed, but we omit the step.
+
+		// Restart Port 1 auto-negotiation
+		register_set(RegisterOffset::P1CR, 1 << 13);
+
+		// Configure Low Watermark to 6KByte available buffer space out of
+		// 12KByte.
+		register_write(RegisterOffset::FCLWR, 0x0600);
+		// Configure High Watermark to 4KByte available buffer space out of
+		// 12KByte.
+		register_write(RegisterOffset::FCHWR, 0x0400);
+
+		// Clear the interrupt status
+		register_write(RegisterOffset::ISR, 0xFFFF);
+		receiveInterruptFutex =
+		  interrupt_futex_get(STATIC_SEALED_VALUE(EthernetInterruptCap));
+		// Enable Receive interrupt
+		register_write(RegisterOffset::IER, ReceiveInterrupt);
+
+		// Enable QMU Transmit.
+		register_set(RegisterOffset::TXCR, 1 << 0);
+		// Enable QMU Receive.
+		register_set(RegisterOffset::RXCR1, 1 << 0);
+	}
+
+	Ksz8851Ethernet(const Ksz8851Ethernet &) = delete;
+	Ksz8851Ethernet(Ksz8851Ethernet &&)      = delete;
+
+	/**
+	 * This device does not have a unique MAC address and so users must provide
+	 * a locally administered MAC address if more than one device is present on
+	 * the same network.
+	 */
+	static constexpr bool has_unique_mac_address()
+	{
+		return false;
+	}
+
+	static constexpr MACAddress mac_address_default()
+	{
+		return {0x3a, 0x30, 0x25, 0x24, 0xfe, 0x7a};
+	}
+
+	void mac_address_set(MACAddress address = mac_address_default())
+	{
+		register_write(RegisterOffset::MARH, (address[0] << 8) | address[1]);
+		register_write(RegisterOffset::MARM, (address[2] << 8) | address[3]);
+		register_write(RegisterOffset::MARL, (address[4] << 8) | address[5]);
+	}
+
+	uint32_t receive_interrupt_value()
+	{
+		return *receiveInterruptFutex;
+	}
+
+	int receive_interrupt_complete(Timeout *timeout,
+	                               uint32_t lastInterruptValue)
+	{
+		// If there are frames to process, do not enter wait.
+		if (frames_to_process)
```

Comment:
Shouldn't we be querying the device for this value, rather than using a cached value that may now be stale here?

Reply:
The device register is only valid after you acknowledge the interrupt. This behaviour is documented in `receive_frame`